src/maps/perf_map_poller.rs

```diff
@@ -0,0 +1,150 @@
+use std::{
+    collections::HashMap,
+    fmt::{self, Formatter},
+    sync::{Arc, Condvar, Mutex},
+    thread,
+    time::Duration,
+};
+
+use crossbeam_channel::{Sender, TrySendError};
+use mio::{unix::SourceFd, Events, Interest, Poll, Token};
+use nix::errno::Errno;
+use slog::crit;
+
+use crate::{
+    maps::{PerCpu, PerfEvent, PerfMap},
+    PerfChannelMessage, LOGGER,
+};
+
+pub struct PerfMapPoller {
+    poll: Poll,
+    tokens: HashMap<Token, PerfMap>,
+}
+
+impl PerfMapPoller {
+    pub fn new(
+        perfmaps: impl Iterator<Item = PerfMap>,
+        polling_signal: Arc<(Mutex<bool>, Condvar)>,
+    ) -> Result<Self, InitError> {
+        let poll = Poll::new().map_err(InitError::Creation)?;
+        let registry = poll.registry();
+
+        let tokens = perfmaps
+            .map(|p| {
+                let token = Token(p.ev_fd as usize);
+                registry
+                    .register(&mut SourceFd(&p.ev_fd), token, Interest::READABLE)
+                    .map(|_| (token, p))
+            })
+            .collect::<Result<_, _>>()
+            .map_err(InitError::Registration)?;
+
+        {
+            // now that the perfmap fd's are registered, we can signal to the main thread that
+            // event polling is ready.
+            let (lock, cvar) = &*polling_signal;
+            let mut locked_signal = lock
+                .lock()
+                .map_err(|e| InitError::ReadySignal(e.to_string()))?;
+            *locked_signal = true;
+            cvar.notify_one();
+        }
+
+        Ok(Self { poll, tokens })
+    }
+
+    pub fn poll(
+        mut self,
+        tx: Sender<PerfChannelMessage>,
+        polling_delay: Duration,
+    ) -> Result<(), std::io::Error> {
+        let mut events = Events::with_capacity(self.tokens.len());
```

Comment:
this mimics what libbpf does and it works in practice: https://github.com/torvalds/linux/blob/master/tools/lib/bpf/libbpf.c#L11162

See how the max events is the number of cpus. In our case it's different because it's number of cpus * number of programs, which equals the number of tokens.